File: 208-implement_trie.py
class TrieNode:
    def __init__(self, c = ""):
        self.c = c
        self.children = {}
        self.is_word = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
        
    def insert(self, word: str) -> None:
        n = self.root
        for c in word:
            if c not in n.children:
                n.children[c] = TrieNode(c)
            n = n.children[c]
        n.is_word = True

    def search(self, word: str) -> bool:
        n = self.root
        for c in word:
            if c not in n.children:
                return False
            n = n.children[c]
        return n.is_word
        
    def startsWith(self, prefix: str) -> bool:
        n = self.root
        for c in prefix:
            if c not in n.children:
                return False
            n = n.children[c]
        return True
        
# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208-implement_trie.cs
public class TrieNode
{
    public char? c;
    public Dictionary<char, TrieNode> children;
    public bool is_word;

    public TrieNode(char c='/') // '/' formally designates root
    {
        this.c = c;
        this.children = new();
        this.is_word = false;
    }
}

public class Trie
{
    private TrieNode root;

    public Trie()
    {
        root = new();
    }
    
    public void Insert(string word)
    {
        var node = this.root;
        foreach(char c in word)
        {
            if (!node.children.ContainsKey(c)) {
                node.children[c] = new TrieNode(c);
            }
            node = node.children[c];
        }
        node.is_word = true;
    }
    
    public bool Search(string word)
    {
        var node = this.root;
        foreach(char c in word)
        {
            if (!node.children.ContainsKey(c)) {
                return false;
            }
            node = node.children[c];
        }
        
        return node.is_word;
    }
    
    public bool StartsWith(string prefix)
    {
        var node = this.root;
        foreach(char c in prefix)
        {
            if (!node.children.ContainsKey(c)) {
                return false;
            }
            node = node.children[c];
        }
        
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.Insert(word);
 * bool param_2 = obj.Search(word);
 * bool param_3 = obj.StartsWith(prefix);
 */
